xml_struct_derive/src/properties.rs

```diff
@@ -0,0 +1,328 @@
+                    }
+                }
+
+                _ => {
+                    errors.push(Error::new(meta.span(), UNRECOGNIZED_ATTRIBUTE_MSG));
+                }
+            }
+        }
+
+        let has_namespace_decl =
+            properties.default_ns_name.is_some() || !properties.ns_decls.is_empty();
+        if has_namespace_decl && properties.should_serialize_as_text {
+            // There's no meaningful way to namespace text content, so the
+            // combination of these properties is almost certainly a mistake.
+            errors.push(Error::new(
+                attr.span(),
+                "cannot declare namespaces for text content",
+            ));
+        }
+
+        if properties.ns_prefix_for_variants.is_some() && properties.should_serialize_as_text {
+            errors.push(Error::new(
+                attr.span(),
+                "cannot declare variant namespace prefix for text enum",
+            ));
+        }
+
+        // Combine and return errors if there are any. If none, we've
+        // successfully parsed the attributes and can return the appropriate
+        // props.
+        match errors.into_iter().reduce(|mut combined, err| {
+            combined.combine(err);
+
+            combined
+        }) {
+            Some(err) => Err(err),
+            None => Ok(properties),
+        }
+    }
+}
+
+#[derive(Debug)]
+/// A declaration of an XML namespace for a type with a derived `XmlSerialize`
+/// implementation.
+pub(crate) struct NamespaceDecl {
+    pub prefix: TokenStream,
+    pub name: TokenStream,
+}
+
+#[derive(Debug, Default)]
+/// Properties governing the serialization of a field in a struct or enum with a
+/// derived `XmlSerialize` implementation.
+pub(crate) struct FieldProps {
+    /// The type of XML structure which the field represents.
+    pub repr: FieldRepr,
+
+    /// Whether the field should be serialized with a "flat" representation.
+    ///
+    /// A flattened field will be serialized only as its content nodes, rather
+    /// than as an XML element containing those content nodes.
+    pub should_flatten: bool,
+
+    /// A prefix to add to this field's name when serialized as an element or
+    /// attribute.
+    pub namespace_prefix: Option<TokenStream>,
+}
+
+impl FieldProps {
+    /// Constructs a set of serialization properties for an enum or struct field
+    /// from its struct attributes.
+    pub(crate) fn try_from_attrs(
+        value: Vec<Attribute>,
+        is_named_field: bool,
+    ) -> Result<Self, Error> {
+        // Find the attribute for configuring behavior of the derivation, if
+        // any.
+        let attr = match value
+            .into_iter()
+            .find(|attr| attr.path().is_ident(MACRO_ATTRIBUTE))
+        {
+            Some(attr) => attr,
+
+            // If we don't find a matching attribute, we assume the default set
+            // of properties.
+            None => return Ok(Self::default()),
+        };
```

Comment:
issue: I find this bit difficult to parse, due to the amount of things happening to define `attr` (chained calls followed by the arms of the match statement, one of which makes the function return and the other doesn't). How about splitting the operation? I find the example bellow more readable but ymmv

```suggestion
        let attr = value
            .into_iter()
            .find(|attr| attr.path().is_ident(MACRO_ATTRIBUTE));

        let attr = match attr {
            Some(attr) => attr,

            // If we don't find a matching attribute, we assume the default set
            // of properties and return immediately.
            None => return Ok(Self::default()),
        };
```

Reply:
I've split the iteration and `find()` call into a helper function to hopefully make this read a little better.